pkg/controller/openapi/openapi_product_reconciler.go

```diff
@@ -0,0 +1,360 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if p.Logger().V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(desired, "", "  ")
+		if err != nil {
+			return nil, err
+		}
+		p.Logger().V(1).Info(string(jsonData))
+	}
+
+	return nil, p.ReconcileResource(&capabilitiesv1beta1.Product{}, desired, p.productMutator)
+}
+
+func (p *OpenAPIProductReconciler) desired() (*capabilitiesv1beta1.Product, error) {
+	// product obj name
+	objName := p.desiredObjName()
+
+	errStrings := validation.IsDNS1123Label(objName)
+	if len(errStrings) > 0 {
+		return nil, errors.New(strings.Join(errStrings, "--"))
+	}
+
+	// product name
+	name := p.openapiObj.Info.Title
+
+	// product system name
+	systemName := p.desiredSystemName()
+
+	// product description
+	description := fmt.Sprintf(p.openapiObj.Info.Description)
+
+	product := &capabilitiesv1beta1.Product{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       capabilitiesv1beta1.ProductKind,
+			APIVersion: capabilitiesv1beta1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objName,
+			Namespace: p.openapiCR.Namespace,
+		},
+		Spec: capabilitiesv1beta1.ProductSpec{
+			Name:               name,
+			SystemName:         systemName,
+			Description:        description,
+			ProviderAccountRef: p.openapiCR.Spec.ProviderAccountRef,
+		},
+	}
+
+	// Deployment
+	product.Spec.Deployment = p.desiredDeployment()
+
+	// Methods
+	product.Spec.Methods = p.desiredMethods()
+
+	// Mapping rules
+	mappingRules, err := p.desiredMappingRules()
+	if err != nil {
+		return nil, err
+	}
+	product.Spec.MappingRules = mappingRules
+
+	// backend usages
+	// current implementation assumes same system name for backend and product
+	backendSystemName := p.desiredSystemName()
+	product.Spec.BackendUsages = map[string]capabilitiesv1beta1.BackendUsageSpec{
+		backendSystemName: capabilitiesv1beta1.BackendUsageSpec{
+			Path: "/",
+		},
+	}
+
+	product.SetDefaults(p.Logger())
+
+	// internal validation
+	validationErrors := product.Validate()
+	if len(validationErrors) > 0 {
+		return nil, errors.New(validationErrors.ToAggregate().Error())
+	}
+
+	err = p.SetOwnerReference(p.openapiCR, product)
+	if err != nil {
+		return nil, err
+	}
+
+	return product, nil
+}
+
+func (p *OpenAPIProductReconciler) productMutator(existingObj, desiredObj common.KubernetesObject) (bool, error) {
+	existing, ok := existingObj.(*capabilitiesv1beta1.Product)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *capabilitiesv1beta1.Product", existingObj)
+	}
+	desired, ok := desiredObj.(*capabilitiesv1beta1.Product)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *capabilitiesv1beta1.Product", desiredObj)
+	}
+
+	// Metadata labels and annotations
+	updated := helper.EnsureObjectMeta(existing, desired)
+
+	// OwnerRefenrence
+	updatedTmp, err := p.EnsureOwnerReference(p.openapiCR, existing)
+	if err != nil {
+		return false, err
+	}
+	updated = updated || updatedTmp
+
+	// Maybe too rough compare method?
```

Comment:
That's indeed a tough question...

Let's leave it like this at the moment and see how people use it and what kind of interactions happen between controllers.